cmd/crypto/helpers.go

```diff
@@ -0,0 +1,84 @@
+/*
+ * MinIO Cloud Storage, (C) 2018 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package crypto
+
+import (
+	"bytes"
+	"crypto/rand"
+	"io"
+	"io/ioutil"
+
+	"github.com/minio/sio"
+	"golang.org/x/crypto/argon2"
+)
+
+// IsEncryptedData - guess if the data is encrypted.
+func IsEncryptedData(data []byte) bool {
+	buf := data[32:]
+	buf = buf[:2]
+	if buf[0] != sio.Version20 {
+		return false
+	}
+	if buf[1] != sio.CHACHA20_POLY1305 {
+		return false
+	}
+	return true
+}
+
+// EncryptData - encrypts data with input password.
+func EncryptData(password string, data []byte) ([]byte, error) {
+	salt := make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
+		return nil, err
+	}
+
+	// derive an encryption key from the master key and the nonce
+	var key [32]byte
+	copy(key[:], argon2.IDKey([]byte(password), salt, 1, 64*1024, 4, 32))
+
+	encrypted, err := sio.EncryptReader(bytes.NewReader(data), sio.Config{
+		Key:          key[:],
+		CipherSuites: []byte{sio.CHACHA20_POLY1305},
+	})
+	if err != nil {
+		return nil, err
+	}
+	edata, err := ioutil.ReadAll(encrypted)
+	return append(salt, edata...), err
+}
+
+// DecryptData - decrypts server config data.
+func DecryptData(password string, data []byte) ([]byte, error) {
+	reader := bytes.NewReader(data)
+	salt := make([]byte, 32)
+	if _, err := io.ReadFull(reader, salt); err != nil {
+		return nil, err
+	}
+	// derive an encryption key from the master key and the nonce
+	var key [32]byte
+	copy(key[:], argon2.IDKey([]byte(password), salt, 1, 64*1024, 4, 32))
```

Comment:
```suggestion
	key := argon2.IDKey([]byte(password), salt, 1, 64*1024, 4, 32)
```

Maybe check length. This way we don't risk a zero key.

Reply:
It is guaranteed that argon2 will always return a key, otherwise, it will panic..

Reply:
As the author of the x/crypto/argon2 package I can confirm that the `IDKey` (as well as `Key`) always return a key of the requested length (`0 <= keyLen <= 2^32 - 1`) - in this case 32 byte (see last argument).
If the `IDKey` function behaves differently this is a bug in the argon2 package.

Reply:
Great. Just get rid of the extra copy then.